Update dependencies

ch1bo opened the pull request:
Updating all kinds of dependencies to latest versions and does refactor bits and pieces to enable this update.

:rowboat: `plutus-ledger-api` module name change. Nothing special, but many places where we import these.

:rowboat: Plutus renamed `ValidatorHash` to `ScriptHash` and there is also no `Script` anymore. Plutus tx code gets compiled directly into a `SerialisedScript

:rowboat: `io-classes` moved symbols around into new module names. Also many places where we import these.
  
:rowboat: `HashableScriptData` in cardano-api instead of normal ScriptData, which keeps original bytes around (when available). This is mostly noise to us as we often create this data from values in our code. Conversions `toScriptData` and `fromScriptData` are using this type now and many places in the code were also updated mechanically. 

:rowboat: Backported the export of `mkTermToEvaluate` to `plutus-ledger-api 1.1.1.0`. Need to wait for `1.2.0.0` support in `cardano-api` et al to not need to do this. 

:rowboat: `cardano-ledger` changed & moved multiple modules and renamed types, mostly via DEPRECATED pragmas :green_heart: The only problem here is that we have been using Ledger code all over the place and we would have been better off by not depending directly onto ledger that much.

:rowboat: Refactored direct chain layer to use `cardano-api` IPC for the node connection, as changes to ouroboros-network{-framework} et al were confusing and I saw the chance to remove a lot of code and work on the `cardano-api` types directly (instead of `consensus` and `ledger` imports).

:rowboat: Incorporated the custom script context work from #777 to make the scripts fit the publish transaction.

:rowboat: Enabled `-Wunused-packages` in all our packages and removed all redundant dependencies. Many things are not required anymore because of the refactoring mentioned above.

:rowboat: Bonus: updated HLS and fourmolu versions in our shell :tada:

---

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

  - `REVIEW` of the byron epoch slots .. should find that out some time, or not as irrelevant?
  - `XXX` Notes on the source-repository packages
  - `XXX` Annoyances with missing exported data constructors (can be addressed in a future update once our PRs are in cardano-api)
  - `XXX` Notes on the ledger code in case we would want to go forward there, that we should use lenses (according to the ledger team)

ch1bo committed:
Bump index states and add s-r-p for cardano-node

The latest cardano-node master seems to be integrating newer versions of
cardano-ledger, plutus et al and so we use it as a starting point.

ch1bo committed:
Update hydra-prelude to support new io-classes

ch1bo committed:
Remove HasPlutusScriptVersion from hydra-cardano-api

Upstream now exported IsPlutusScriptLanguage, which we can use instead.

ch1bo committed:
Fix / update UsingRawBytesHex class/instances

This should be rather taken from upstream. Created two PRs to avoid
workarounds: https://github.com/input-output-hk/cardano-node/pull/5085
and https://github.com/input-output-hk/cardano-node/pull/5086

ch1bo committed:
Fix imports from plutus-ledger-api

The V1/V2 plutus ledger api module names have changed.

ch1bo committed:
Use HashableScriptData in conversion functions

This new type was introduced to improve roundtrip equality of script
data to/from the ledger. Unsure which version we should need when though.

ch1bo committed:
Drop orphan instances for ChainPoint and ByronAddr

They seem to have been added to cardano-api upstream.

ch1bo committed:
Upstreamed ScriptWitnessInCtx

The same class and instances are now in cardano-api

ch1bo committed:
Fix usage of ledger types

Seems like the ledger ValidatedTx got renamed to AlonzoTx, paired with a
deprecation notice to not use the type alias.

The ledger team suggests to use the Core.Tx (et al) type aliases and
lenses to create / access era-specific ledger types.

ch1bo committed:
Update usage of Plutus.Script and Ledger.AlonzoScript

Both, plutus-ledger-api and cardano-ledger changed their interface types.

ch1bo committed:
Avoid using removed "HasField" instances

The "minted" view on the TxBody has been removed. Let's avoid dealing
with the TxBody directly by using 'rdptr' function from the
ledger (spec).

ch1bo committed:
Use MaryValue instead of the deprecated ledger Value type

ch1bo committed:
Rename getScriptData -> txOutScriptData and change signatures

Fixing compilation of the Cardano.Api.Pretty module showed that the
HashableScriptData type is more general and likely better suited for users.

ch1bo committed:
Move genTxIn and Arbitrary TxIn to hydra-cardano-api

We use it also in hydra-test-utils and these generators could all live
in hydra-cardano-api for greater re-usability.

ch1bo committed:
Greatly simplify hydra-test-utils

There is still a big overlap between this and the utilities in
hydra-node, but using cardano-api types makes the redundancy super clear
to refactor later. Moving away from the ledger types also made it
simpler / less annoying to update to new / changed types when updating dependencies.

ch1bo committed:
Update plutus-cbor to new types & interfaces

ch1bo committed:
Update plutus-merkle-tree to new types & interfaces

ch1bo committed:
Use standard cabal boilerplate also hydra-test-utils

ch1bo committed:
Update haskell.nix, CHaP and disable some packages

This enables the "normal" devShell again to make fixing our packages
easier (HLS, hoogle and fourmolu).

ch1bo committed:
Update haskell-language-server and fourmolu in devShell

With latest haskell.nix and a workaround from cardano-node repository we
can update to the latest haskell-language-server and fourmolu versions
in our default development shell:

- haskell-language-server 1.10.0.0
- fourmolu 0.9.0.0

ch1bo committed:
Resolve warnings in MarkdownFormatterSpec

ch1bo committed:
Fix imports and update types in hydra-plutus

ch1bo committed:
Start updating dependencies for hydra-node

This makes pinned deps as lower bounds and enable the package.

ch1bo committed:
Fix usage of io-classes in Logging

ch1bo committed:
Update ouroboros-network-framework usage sites

The exact semantics of this MakeBearer are not clear to me though.

ch1bo committed:
Fix usage of makeTransactionBodyAutoBalance

The cardano-api seems to be moving into the "EpochInfo" types of
consensus

ch1bo committed:
Make Hydra.Ledger.Cardano.Evaluate compile

This required disabling prepareTxScripts function as the
mkTermToEvaluate function is only exported from plutus-ledger-api
version 1.2.0.0 onwards. However, that version is already incompatible
with the upper bounds set in cardano-api et al.

ch1bo committed:
Rewrite Hydra.Chain.Direct to use cardano-api IPC

Could not figure out how to fix the NodeToClient ouroboros stuff
otherwise. This should also unclutter things a bit and hopefully also
works (will only know when everything compiles).

ch1bo committed:
Fix JSON instances for ledger types

This required some more / different type class constraints due to
internal changes in the ledger.

ch1bo committed:
Move genTxIn and fix imports in Hydra.Ledger.Cardano

ch1bo committed:
Fix compilation of several Hydra modules

Some instances are incomplete as it is missing the
SerialiseAsRawBytesError from cardano-api to produce the proper Left
values. See https://github.com/input-output-hk/cardano-node/pull/5085

ch1bo committed:
Fix imports in Hydra.Chain.Direct.Wallet

This is only a makeshift and the import list really needs some
treatment. Ideally we would shift away from using ledger types directly.

ch1bo committed:
Fix io-classes imports in hydra-node

ch1bo committed:
Introduce fromScriptData and use in Hydra.Chain.Direct.Tx

ch1bo committed:
Fix ledger type imports in Handlers

ch1bo committed:
Add getChainPoint to hydra-cardano-api

ch1bo committed:
Complete refactor of Hydra.Chain.Direct modules

Now we are using cardano-api types in the chain sync and local tx
submission clients, as well as in the callbacks to the wallet. The
wallet internals still use quite a lot ledger types.

ch1bo committed:
Fix compilation of tx-cost

ch1bo committed:
Fix conversion of test cost models

For some reason the cardano-ledger-alonzo-test package was zeroing out
all cost model parameters.

ch1bo committed:
Use new MonadSTM exports in hydra-node tests

ch1bo committed:
Add BlockHeader generator and Arbitrary instance to hydra-cardano-api

ch1bo committed:
Change ChainSyncHandlers onRollForward to take a BlockHeader + [Tx]

This is sufficient and does not require data constructors from the
cardano-api type Block.

ch1bo committed:
Change Wallet update to take BlockHeader and [Tx]

This matches the onRollForward and makes WalletSpec tests slightly
simpler. However the Wallet(Spec) still use a lot of ledger types and
require major refactor.

ch1bo committed:
Fix compilation errors in mutation tests

ch1bo committed:
Fix CryptoSpec

The new cardano-base packed bytes prevent making too short signatures. Nice.

ch1bo committed:
Update model based tests to quickcheck-dynamic 3.x

This required to put some (orphan) instances to avoid deriving Generic
on Party type because of the new HasVariables mechanism (which we do
not seem to use).

ch1bo committed:
Script envelope decoding does not support SimpleScriptV2 anymore

Seemingly the decoder changed upstream.

ch1bo committed:
Remove a redundant new line on a golden file

Seemingly the golden file tests have gotten sensitive to this. Renaming
the .faulty.reencoded.json file as the original golden file resolved
this.

ch1bo committed:
Make hydra-cluster, hydra-tui and hydraw compile

ch1bo committed:
Link node and chain threads in ModelSpec

This makes tests fail should an exception be raised on the (mock) chain
component or the logic event processing threads.

ch1bo committed:
Create blocks with increasing slots in MockChain

This has the Model tests pass as the arbitrary TimeHandle is having a
"high enough" horizon to convert these. However, we should not use an
arbitrary time handle, but one that is more realistically oriented from
a "current slot" (which is kept now in a TVar in MockChain).

ch1bo committed:
Fix warning in Hydra.Ledger.Cardano.Builder

ch1bo committed:
Fix warnings and remove redundant dependencies in hydra-node

ch1bo committed:
Address warnings in hydra-cardano-api

ch1bo committed:
Address warnings in hydra-cluster and hydraw

ch1bo committed:
Always build benchmarks

Similar to tests, we should make sure that benchmarks compile. By
building them always (for local packages) we are reminded more easily.

ch1bo committed:
Change wallet sign & coverFee functions to use cardano-api Tx

This is another step in getting rid of ledger types in our code base.
The cardano-api types are higher level and a more stable api. Hence we
prefer to use those types.

v0d1ch commented:
I added a small commit to improve on getting the free port. After this I am noticing that cluster tests are hanging and in the cardano-node log this:
```
{"app":[],"at":"2023-04-19T11:55:17.61Z","data":{"error":"user interrupt","kind":"ConsensusStartupException"},"env":"1.35.7:f0b4a","host":"nixos","loc":null,"msg":"","ns":["cardano.node"],"pid":"376200","sev":"Critical","thread":"5"}

```

ch1bo committed:
Bump hackage to get newer quickcheck-dynamic

This also required adding a constraint to avoid using a too new
unix-bytestring in ouroboros-consensus and bumping CHaP was adding
some other integration / compilation errors. The constraint is easier
for now.

ch1bo committed:
Use forAllNonVariableQ and remove an orphan instance

We can now quantify without "HasVariables"

ch1bo committed:
Re-enable prepareTxScripts using backported export of mkTermToEvaluate

ffakenz commented on hydra-cardano-api/src/Hydra/Cardano/Api/ScriptData.hs:
```diff
@@ -89,3 +95,14 @@ instance FromJSON ScriptData where
       bytes <- Base16.decode (encodeUtf8 text)
       left show $ deserialiseOrFail $ fromStrict bytes
+
+-- NOTE: We are okay with storing the data and re-serializing + hasing it into
```
minor typo
```suggestion
-- NOTE: We are okay with storing the data and re-serializing + hashing it into
```

ffakenz commented on hydra-node/hydra-node.cabal:
```diff
@@ -63,7 +63,7 @@ common project-config
   ghc-options:
     -Wall -Wcompat -Widentities -Wincomplete-record-updates
-    -Wincomplete-uni-patterns -Wredundant-constraints
+    -Wincomplete-uni-patterns -Wredundant-constraints -Wunused-packages
```
are we sure about keeping this?

ch1bo replied:
Yes, until it is too annoying. We had too many dependencies and we would've removed them earlier if we have this enabled.

ffakenz commented on plutus-cbor/plutus-cbor.cabal:
```diff
@@ -85,7 +85,7 @@ test-suite unit
   hs-source-dirs:     test
   main-is:            Main.hs
   ghc-options:        -threaded -rtsopts
-  build-tool-depends: hspec-discover:hspec-discover -any
```
what is this -any argument for and why are we removing this?

ch1bo replied:
It's a version indicator I guess. My formatter is removing it, and other peoples formatter is adding it .. or the other way around. It doesn't change anything if we have it or not.

ffakenz reviewed (approved):
Massive effort! Nice job 👏 

![image](https://user-images.githubusercontent.com/14299604/233726066-0a1f512d-e15c-4d10-82b0-aab2c016d4c1.png)

ch1bo committed:
Regenerate golden hydra-plutus script files

ch1bo committed:
Fix a typo in hydra-cardano-api haddocks

Co-authored-by: Franco Testagrossa <franco.testagrossa@iohk.io>

v0d1ch reviewed (approved):
Nice effort! You checked the _no new TODO's added_ checkbox but I still see some (like: `TODO: Use IsPlutusScriptLanguage here` so not sure if you wanted to address those before merging or not.

ch1bo committed:
Clarify a TODO as a NOTE

ch1bo committed:
Use IsPlutusScriptLanguage for a more generic orphan instance

ch1bo committed:
Make hydra-cardano-api import explicit

We favor explicit imports for better reviewability.

ch1bo committed:
Remove the unbounded 'eraHistory' from fixtures

We can use the lower level and "smaller type" of 'EpochInfo' now with
the cardano-api. So there is no need to keep that fixture anymore.

The 'TimeHandle' usages and tests still need the
'eraHistoryWithHorizonAt' (for now).

ch1bo committed:
Hack E2E tests: wait for a bit before restarting Bob's node

Seems like due to recent updates in the ouroboros-network package and/or
the way we use that networking stack, the threads don't come down
cleanly / instantly. Have not found a way to properly locate, but the
network port seems available after a little while.

ch1bo commented:
> Nice effort! You checked the _no new TODO's added_ checkbox but I still see some (like: `TODO: Use IsPlutusScriptLanguage here` so not sure if you wanted to address those before merging or not.

@v0d1ch resolved all `TODO`s now.

Needed to introduce a `HACK` to make the `hydra-cluster` tests pass though :/ https://github.com/input-output-hk/hydra/commit/fc7f08bfa6b6fe18533ce1be40a5dcc584461855
